# 09-statement-metrics

### 1. Загрузите файл classification.csv. В нем записаны истинные классы объектов выборки (колонка true) и ответы некоторого классификатора (колонка predicted).

In [1]:
import pandas as pd

url_class = 'https://raw.githubusercontent.com/Sazankova/metrics/main/classification.csv'
data = pd.read_csv(url_class)

### 2. Заполните таблицу ошибок классификации: Для этого подсчитайте величины TP, FP, FN и TN согласно их определениям. Например, FP это количество объектов, имеющих класс 0, но отнесенных алгоритмом к классу1. Ответ в данном вопросе четыре числа через пробел.

In [2]:
tp, fp, fn, tn = 0, 0, 0, 0

for i in range(len(data)):
    if data['true'][i] == 1 and data['pred'][i] == 1:
        tp += 1
    elif data['true'][i] == 0 and data['pred'][i] == 1:
        fp += 1
    elif data['true'][i] == 1 and data['pred'][i] == 0:
        fn += 1
    elif data['true'][i] == 0 and data['pred'][i] == 0:
        tn += 1

print(tp, fp, fn, tn)


43 34 59 64


### 3. Посчитайте основные метрики качества классификатора: • Accuracy (доля верно угаданных) sklearn.metrics.accuracy • Precision (точность) sklearn.metrics.accuracy.precision_score • Recall (полнота) sklearn.metrics.recall_score • F-мера sklearn.metrics.f1_score

In [3]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_true = data['true']
y_pred = data['pred']

accuracy = accuracy_score(y_true, y_pred)
precision = precision_score(y_true, y_pred)
recall = recall_score(y_true, y_pred)
f1 = f1_score(y_true, y_pred)

print(accuracy, precision, recall, f1)

0.535 0.5584415584415584 0.4215686274509804 0.48044692737430167


### 4. Имеется четыре обученных классификатора. В файле scores.csv записаны истинные классы и значения степени принадлежности положительному классу для каждого классификатора на некоторой выборке: • для логистической регрессии вероятность положительного класса (колонка score_logreg), • дляSVM отступотразделяющейповерхности(колонкаscore_svm), • для метрического алгоритма взвешенная сумма классов соседей (колонка score_knn), • для решающего дерева доля положительных объектов в листе (колонка score_tree). Загрузите этот файл.

In [4]:
url_scores = 'https://raw.githubusercontent.com/Sazankova/metrics/main/scores.csv'
data_score = pd.read_csv(url_scores)

### 5. Посчитайте площадь под ROC-кривой для каждого классификатора. Какой классификатор имеет наибольшее значение метрики AUC-ROC (укажите название столбца с ответами этого классификатора)? Воспользуйтесь функцией sklearn.metrics.roc_auc_score.

In [5]:
from sklearn.metrics import roc_auc_score

roc_auc_logreg = roc_auc_score(data_score['true'], data_score['score_logreg'])
roc_auc_svm = roc_auc_score(data_score['true'], data_score['score_svm'])
roc_auc_knn = roc_auc_score(data_score['true'], data_score['score_knn'])
roc_auc_tree = roc_auc_score(data_score['true'], data_score['score_tree'])

print(roc_auc_logreg, roc_auc_svm, roc_auc_knn, roc_auc_tree)
print(f"Наибольший AUC-ROC: score_logreg ({roc_auc_logreg:.3f})")

0.719187675070028 0.7086834733893557 0.6351540616246498 0.6919267707082833
Наибольший AUC-ROC: score_logreg (0.719)


### logreg 0.719

### 6. Какой классификатор достигает наибольшей точности (Precision) при полноте (Recall) не менее 70% (укажите название столбца с ответами этого классификатора)? Какое значение точности при этом получается?

In [6]:
from sklearn.metrics import precision_recall_curve

def max_precision_at_recall(y_true, y_scores, min_recall=0.7):
    precision, recall, thresholds = precision_recall_curve(y_true, y_scores)
    max_prec = 0
    for i in range(len(precision) - 1):
        if recall[i] >= min_recall:
            max_prec = max(precision[i], max_prec)
    return max_prec

print('score_logreg:', max_precision_at_recall(data_score['true'], data_score['score_logreg']))
print('score_svm:', max_precision_at_recall(data_score['true'], data_score['score_svm']))
print('score_knn:', max_precision_at_recall(data_score['true'], data_score['score_knn']))
print('score_tree:', max_precision_at_recall(data_score['true'], data_score['score_tree']))

score_logreg: 0.6302521008403361
score_svm: 0.6228070175438597
score_knn: 0.6065573770491803
score_tree: 0.6517857142857143


### tree 0.652